packages/vite/src/node/server/persistentCache.ts

```diff
@@ -0,0 +1,443 @@
+import fs from 'node:fs'
+import path from 'node:path'
+import colors from 'picocolors'
+import type { DepOptimizationMetadata } from '../index'
+import type {
+  InlineConfig,
+  ResolvedConfig,
+  ResolvedServerPersistentCacheOptions
+} from '../config'
+import { normalizePath, version } from '../publicUtils'
+import { createDebugger, getCodeHash, isDefined, lookupFile } from '../utils'
+import type { ModuleNode } from './moduleGraph'
+
+export interface PersistentCache {
+  manifest: PersistentCacheManifest
+  getKey: (code: string) => string
+  read: (key: string) => Promise<PersistentCacheResult | null>
+  write: (
+    key: string,
+    id: string,
+    url: string | undefined,
+    mod: ModuleNode | null,
+    ssr: boolean,
+    file: string,
+    code: string,
+    map?: any
+  ) => Promise<void>
+  queueManifestWrite: () => void
+  updateDepsMetadata: (metadata: DepOptimizationMetadata) => Promise<void>
+}
+
+export interface PersistentCacheManifest {
+  version: string
+  modules: Record<string, PersistentCacheEntry>
+  files: Record<string, PersistentCacheFile>
+}
+
+export interface PersistentSimpleCacheEntry {
+  id: string
+  url?: string
+  file: string
+  fileCode: string
+  fileMap?: string
+}
+
+export interface PersistentFullCacheEntry extends PersistentSimpleCacheEntry {
+  importedModules: { id: string; file: string; url: string }[]
+  importedBindings: Record<string, string[]>
+  acceptedHmrDeps: string[]
+  acceptedHmrExports: string[]
+  isSelfAccepting?: boolean
+  ssr: boolean
+}
+
+export type PersistentCacheEntry =
+  | PersistentSimpleCacheEntry
+  | PersistentFullCacheEntry
+
+export function isFullCacheEntry(
+  entry: PersistentCacheEntry
+): entry is PersistentFullCacheEntry {
+  return Array.isArray((entry as PersistentFullCacheEntry).importedModules)
+}
+
+export interface PersistentCacheResult {
+  code: string
+  map?: any
+}
+
+export interface PersistentCacheFile {
+  relatedModules: Record<string, string>
+}
+
+export async function createPersistentCache(
```

Comment:
just a personal style thing, but I think this method's hard to read because of the various methods defined within it. maybe it'd be better in vs code or something and it's just cause i'm reading it on github, but it's difficult to see where this method begins and ends since it's 300 lines long and I think it'd be clearer to move some of the supporting methods out if possible